### ISSUE TITLE
APOLLO optimizer

### ISSUE THREAD

**Issue by fblgit** [axolotl-ai-cloud/axolotl#2175]
### ⚠️ Please check that this feature request hasn't been suggested before.

- [X] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [X] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Any plan or timeline for APOLLO optimizer?

https://arxiv.org/abs/2412.05270

looks like.. very interesting

### ✔️ Solution

Implementing https://arxiv.org/abs/2412.05270

### ❓ Alternatives

buy me 4x gpu :D

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this feature has not been requested yet.
- [X] I have provided enough information for the maintainers to understand and evaluate this request.

**Comment by winglian** [comment]
Thanks @fblgit! We'll look into this and see how feasible this is without a reference implementation

**Comment by fizzAI** [comment]
https://github.com/zhuhanqing/APOLLO/tree/main/apollo_torch there's a reference implementation now
would be cool to see :3

**Comment by ehartford** [comment]
FYI implementing an optimizer is quite independent from Axolotl.  It's a huggingface transformers thing. 

Check out how I implemented grokAdamW

https://github.com/huggingface/transformers/pull/32521

**Comment by fizzAI** [comment]
Yes, but axo already has some optimizers of its own that it patches into TF's on its own, ie [ADOPT](https://github.com/axolotl-ai-cloud/axolotl/blob/main/src/axolotl/utils/optimizers/adopt.py), and it's probably much easier to get a PR doing that working here than dealing with upstream transformers

**Comment by zhuhanqing** [comment]
@fblgit @fizzAI Thank you for raising a request to add APOLLO! 
@winglian I am the author of  [APOLLO: SGD-like Memory, AdamW-level Performance](https://arxiv.org/abs/2412.05270). Our optimizer has been integrated into [LLaMA-Factory](https://github.com/hiyouga/LLaMA-Factory/pull/6617), [FluxML](https://github.com/FluxML/Optimisers.jl/pull/196), and with a validated performance in the [post](https://bsky.app/profile/benjmurrell.bsky.social/post/3lcyfrf5b7k2u). We are also undergoing integration into [Huggingface Transformers](https://github.com/huggingface/transformers/pull/36062).

We use SGD-like memory cost but can achieve on-par or even better performance than AdamW, validated on both pre-training and fine-tuning (even pre-training a LLaMA-7B model). Please let us know if you want to integrate APOLLO and have any issues during integration!

**Comment by fizzAI** [comment]
@zhuhanqing hai!  
whenever the transformers pr gets pulled, apollo *should* be supported in axo too because it pulls most of its optims from tf

**Comment by fizzAI** [comment]
This issue can be closed now that the transformers PR is merged :) it just works ootb
